crates/store/re_dataframe2/src/query.rs

```diff
@@ -0,0 +1,644 @@
+                let cursor_value = cur_cursor.load(Ordering::Relaxed) as usize;
+
+                // TODO(cmc): This can easily be optimized by looking ahead and breaking as soon as chunks
+                // stop overlapping.
+
+                let Some(cur_row_id) = cur_chunk.row_ids().nth(cursor_value) else {
+                    continue;
+                };
+
+                let Some(cur_index_value) = cur_chunk
+                    .timelines()
+                    .get(&self.query.filtered_index)
+                    .map_or(Some(TimeInt::STATIC), |time_column| {
+                        time_column
+                            .times_raw()
+                            .get(cursor_value)
+                            .copied()
+                            .map(TimeInt::new_temporal)
+                    })
+                else {
+                    continue;
+                };
+
+                streaming_state_per_component
+                    .entry(*component_name)
+                    .and_modify(|streaming_state| {
+                        let StreamingJoinState {
+                            chunk,
+                            cursor,
+                            index_value,
+                            row_id,
+                        } = streaming_state;
+
+                        let cur_chunk_has_smaller_index_value = cur_index_value < *index_value;
+                        // If these two chunks overlap and share the index value of the current
+                        // iteration, we shall pick the row with the most recent row-id.
+                        let cur_chunk_has_equal_index_but_higher_rowid =
+                            cur_index_value == *index_value && cur_row_id > *row_id;
+
+                        if cur_chunk_has_smaller_index_value
+                            || cur_chunk_has_equal_index_but_higher_rowid
+                        {
+                            *chunk = chunk;
+                            *cursor = cursor;
+                            *index_value = cur_index_value;
+                            *row_id = cur_row_id;
+                        }
+                    })
+                    .or_insert_with(|| StreamingJoinState {
+                        chunk: cur_chunk,
+                        cursor: cur_cursor,
+                        index_value: cur_index_value,
+                        row_id: cur_row_id,
+                    });
+            }
+        }
+
+        // What's the index value we're looking for at the current iteration?
+        let cur_index_value = streaming_state_per_component
+            .values()
+            // NOTE: We're purposefully ignoring RowId-related semantics here: we just want to know
+            // the value we're looking for on the "main" index (dedupe semantics).
+            .min_by_key(|streaming_state| streaming_state.index_value)
+            .map(|streaming_state| streaming_state.index_value)?;
```

Comment:
Rather than doing this on every call to next_row, I suspect it might be clearer to do this whole thing as a 2-phased process.

First, work just with the Timeline data from every view-relevant chunk to materialize a new column of sorted/unique TimeInt values (note, as an added benefit this is the same input you'll want to be able to feed into `sample_index_values()` anyways).  This could still be done incrementally, "batch-wise" by only looking at overlapping chunks on some horizon. 

Then, once we have the ability to iterate over batches of TimeInts, we iterate through them incrementally and look for the matching values from the relevant chunks, as you're doing below, which now becomes a common code-path between this implementation and `sampled_index_values()`

Additionally, my gut is that having batches of unique TimeInts in advance sets us up nicely for some future optimizations.
- It lets us fairly easily parallelize the per-selected-column work. Each worker can independently yield a sequence of rows matching the requested sequence of TimeInts.
- It lets us look ahead to check for matching runs in the given columns. Any time we have a matching run in a range with a single column (happy path) we can directly yield a slice of multiple rows from our column-generator.
- Similarly, null runs can quickly be identified and generated when the last TimeInt in the requested batch is less than the next available time-int for the column.
- The aggregator consuming from each of the parallel columns generators can then yield RecordBatches based on overlapping row-runs from the separate columns, which means in the happy path of dense non-overlapping chunks we return to getting nice contiguous slices again.

Reply:
We can do these improvements in follow up PRs, let's focus on landing all semantics first.

Reply:
Agreed -- not a requested change. Just an observation about the structure to keep in mind as you refactor in the direction of supporting `sampled_index_values()`